fv/sockmap_test.go

```diff
@@ -0,0 +1,322 @@
+			prepend0xInPlace(fields)
+			entry := mapEntry{
+				Key: fields,
+			}
+			al = append(al, entry)
+		}
+		phase = (phase + 1) % 4
+	}
+}
+
+func getSockmapOpts() infrastructure.TopologyOptions {
+	opts := infrastructure.DefaultTopologyOptions()
+	opts.EnableIPv6 = false
+	opts.FelixLogSeverity = "debug"
+	opts.ExtraEnvVars["FELIX_XDPENABLED"] = "0"
+	opts.ExtraEnvVars["FELIX_SOCKMAPENABLED"] = "1"
+	return opts
+}
+
+var _ = infrastructure.DatastoreDescribe("with Felix using sockmap", []apiconfig.DatastoreType{apiconfig.EtcdV3 /*, apiconfig.Kubernetes*/}, func(getInfra infrastructure.InfraFactory) {
+	var (
+		infra   infrastructure.DatastoreInfra
+		felix   *infrastructure.Felix
+		host    *workload.Workload
+		ip      string
+		port    int
+		srcPort int
+	)
+
+	BeforeEach(func() {
+		if err := bpf.SupportsSockmap(); err != nil {
+			Skip(fmt.Sprintf("Sockmap acceleration not supported: %v", err))
+		}
+		infra = getInfra()
+		opts := getSockmapOpts()
+		felix, _ = infrastructure.StartSingleNodeTopology(opts, infra)
+		ip = "10.65.0.2"
+		port = 8055
+		srcPort = 8056
+		host = workload.Run(
+			felix,
+			"service",
+			"default",
+			ip,
+			fmt.Sprintf("%d", port),
+			"tcp")
+		host.ConfigureInDatastore(infra)
+	})
+
+	AfterEach(func() {
+		host.Stop()
+		felix.Stop()
+		func() {
+			opts := getSockmapOpts()
+			felix, _ = infrastructure.StartSingleNodeTopology(opts, infra)
+			defer felix.Stop()
+			outputBytes, err := utils.Command(
+				"docker",
+				"exec",
+				felix.Container.Name,
+				"bpftool",
+				"--json",
+				"--pretty",
+				"map",
+				"dump",
+				"pinned",
+				"/sys/fs/bpf/calico/sockmap/calico_sock_map_v1",
+			).CombinedOutput()
+			if err != nil {
+				log.WithFields(log.Fields{
+					"containerID": felix.Container.Name,
+					"output":      string(outputBytes),
+				}).WithError(err).Info("Failed to dump the contents of the sock map, skipping cleanup")
+				return
+			}
+			if strings.TrimSpace(string(outputBytes)) != "[]" {
+				log.WithFields(log.Fields{
+					"containerID": felix.Container.Name,
+					"output":      string(outputBytes),
+				}).Info("Sock map is not empty, skipping cleanup")
+				return
+			}
+			fullCgroupDir := "/run/calico/cgroup"
+			outputBytes, err = utils.Command(
+				"docker",
+				"exec",
+				felix.Container.Name,
+				"bpftool",
+				"cgroup",
+				"detach",
+				fullCgroupDir,
+				"sock_ops",
+				"pinned",
+				"/sys/fs/bpf/calico/sockmap/calico_sockops_v1",
+			).CombinedOutput()
+			if err != nil {
+				log.WithFields(log.Fields{
+					"cgroupdir":   fullCgroupDir,
+					"containerID": felix.Container.Name,
+					"output":      string(outputBytes),
+				}).WithError(err).Info("Failed to detach sockops program from cgroup, skipping cleanup")
+				return
+			}
+			log.Info("Cleanup finished")
+		}()
+		infra.Stop()
+	})
+
+	It("should put the IP of the host in sockmap endpoints map", func() {
+		found := false
+		maxTries := 5
+		for try := 0; try < maxTries; try++ {
```

Comment:
Ginkgo has quite a nice way of doing retry loops.  If you made the function return the `entries` slice, you could do this:
```
Eventually(getEntries, "5s", "100ms").Should(Contain(expectedSlice))
```

Reply:
Yeah, I rewrote it to use `Eventually`.